src/v/kafka/server/group_data_parser.h

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * as of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "features/feature_table.h"
+#include "kafka/server/group.h"
+#include "kafka/server/group_metadata.h"
+#include "kafka/server/logger.h"
+#include "model/record.h"
+
+template<typename T>
+T parse_tx_batch(const model::record_batch& batch, int8_t version) {
+    vassert(batch.record_count() == 1, "tx batch must contain a single record");
+    auto r = batch.copy_records();
+    auto& record = *r.begin();
+    auto key_buf = record.release_key();
+    auto val_buf = record.release_value();
+
+    iobuf_parser val_reader(std::move(val_buf));
+    auto tx_version = reflection::adl<int8_t>{}.from(val_reader);
+    vassert(
+      tx_version == version,
+      "unknown group inflight tx record version: {} expected: {}",
+      tx_version,
+      version);
+    auto cmd = reflection::adl<T>{}.from(val_reader);
+
+    iobuf_parser key_reader(std::move(key_buf));
+    auto batch_type = reflection::adl<model::record_batch_type>{}.from(
+      key_reader);
+    const auto& hdr = batch.header();
+    vassert(
+      hdr.type == batch_type,
+      "broken tx group message. expected batch type {} got: {}",
+      hdr.type,
+      batch_type);
+    auto p_id = model::producer_id(reflection::adl<int64_t>{}.from(key_reader));
+    auto bid = model::batch_identity::from(hdr);
+    vassert(
+      p_id == bid.pid.id,
+      "broken tx group message. expected pid/id {} got: {}",
+      bid.pid.id,
+      p_id);
+    return cmd;
+}
+
+namespace kafka {
+template<class Base>
+class group_data_parser {
+protected:
+    ss::future<> parse(model::record_batch b) {
```

Comment:
it looks to me that we do not have to return a `future` from all the handle methods ?

Reply:
its needed for handle_raft_data but not for the rest of them, I just kept it consistent as a 'future' proof incase some 'future' code wants to do something interesting than adding to some in memory state.